### PR TITLE
Convert Bureau Structure to Wagtail

### PR DESCRIPTION
**Pull request by sebworks**
Convert Bureau Structure to Wagtail

## Additions

- Added CSS file to customize Wagtail Admin for the Bureau Structure.

## Changes

- Modified Jinja, Less, and python files to migrate the bureau structure page from a Sheer page to Wagtail. I left the current Sheer Page in place and just updated the necessary associated files to make it atomic.

## Testing
- Run `gulp build`
- Add a Browse page via Wagtail.
- Add a Bureau Structure block to the page
- Verify that the page previews as expected. 

## Review

- @cfpb/cfgov-backends 
- @Scotchester 

## Notes

- It might make sense to use snippets at some point but it would have to more flexible then just injecting that snippet into a 25 / 75 organism

- The CSS modifications to the Wagtail admin are experimental and will probably need to be adjusted.

## Todos

-

## Checklist

* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

### CODE REVIEW

**Comment by sebworks** [comment]
@richaagarwal, can you take a look at this PR.

**Comment by richaagarwal** [comment]
Sure, on it!

**Comment by Scotchester** [comment]
I'll try to review this today, too.

**Comment by richaagarwal** [comment]
Did a quick review locally -- one thing I noticed is that an image is required to upload, but then that image doesn't display.  Not sure how that's supposed to work.

File: cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/_macro-content-slider.html [hunk 1]
```diff
@@ -13,7 +13,7 @@
    ========================================================================== #}
 {% macro render(categories) -%}
 <div id="content-slider">
-    <div class="org-chart_categories">
+    <div class="o-bureau-structure_categories">
```
Comment (richaagarwal) [inline cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/_macro-content-slider.html]: Minor thing but is there a reason all of these names need to be changed?

File: cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/index.html [hunk 2]
```diff
@@ -46,9 +46,9 @@
 
 {% block content_main %}
     <h1>Bureau Structure</h1>
-    <p>Org chart last updated: August 29, 2016</p>
-    <div class="org-chart">
-        <div class="org-chart_node org-chart_node__root">
+    <p>Org chart last updated: November 8, 2016</p>
```
Comment (richaagarwal) [inline cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/index.html]: This file will be deprecated, right?

File: cfgov/v1/atomic_elements/organisms.py [hunk 3]
```diff
@@ -170,6 +170,46 @@ class Meta:
         label = ' '
 
 
+class BureaStructurePosition(blocks.StructBlock):
+    name = blocks.CharBlock()
+    title = blocks.StructBlock([
+        ('line_1', blocks.CharBlock(required=False)),
```
Comment (richaagarwal) [inline cfgov/v1/atomic_elements/organisms.py]: What are Lines 1 and 2 meant to represent? (Maybe this doesn't matter if it's clear to content editors, just curious)

**Comment by richaagarwal** [comment]
Left a few more comments, but otherwise looks good.  I won't be around for the next 10 days so don't wait on me for final approval.  Curious what next steps are -- will the current bureau page be deleted after this is merged in and someone has created the page in Wagtail?

File: cfgov/templates/wagtailadmin/css/bureau-structure.css [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  .blockname-bureau_structure li[id*="office_of_the_director"],
+  .blockname-bureau_structure li[id*="divisions"] {
+    border: 1px solid #e6e6e6 !important;
+    margin: 5px 5px 5px 10px;
+    padding: 10px;
+    border-radius: 6px;
+    min-width: 440px;
+  }
+}
```
Comment (Scotchester) [inline cfgov/templates/wagtailadmin/css/bureau-structure.css]: In general, I think these customizations are fine for now. We should make a more concerted effort to improve the styling of the Wagtail editor at some point. Lots of questionable decisions in there (especially the decreasing widths as you go deeper nested fields).

File: cfgov/v1/atomic_elements/organisms.py [hunk 5]
```diff
@@ -170,6 +170,46 @@ class Meta:
...
+    positions = blocks.ListBlock(BureaStructurePosition(required=False))
+
+
+class BureaStructureOffice(blocks.StructBlock):
+    name = blocks.CharBlock()
+    title = blocks.StructBlock([
+        ('line_1', blocks.CharBlock(required=False)),
+        ('line_2', blocks.CharBlock(required=False))
+    ])
+    office_title = blocks.CharBlock(required=False)
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: Also, I think here it should probably be `division_name`. "Title" could potentially get confused with someone's job title.

File: cfgov/v1/atomic_elements/organisms.py [hunk 6]
```diff
@@ -170,6 +170,46 @@ class Meta:
...
+class BureaStructurePosition(blocks.StructBlock):
+    name = blocks.CharBlock()
+    title = blocks.StructBlock([
+        ('line_1', blocks.CharBlock(required=False)),
+        ('line_2', blocks.CharBlock(required=False))
+    ])
+
+
+class BureaStructureNode(BureaStructurePosition):
+    office_title = blocks.CharBlock(required=False)
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: Suggest calling it `office_name` here, for the same reason mentioned earlier.

File: cfgov/v1/atomic_elements/organisms.py [hunk 7]
```diff
@@ -170,6 +170,46 @@ class Meta:
...
+    ])
+    office_title = blocks.CharBlock(required=False)
+    more_info_url = atoms.Hyperlink(required=False)
+    positions = blocks.ListBlock(BureaStructurePosition(required=False))
+
+
+class BureaStructure(blocks.StructBlock):
+    director = blocks.CharBlock()
+    last_updated_date = blocks.DateBlock(required=False)
+    download_image = images_blocks.ImageChooserBlock(icon='image')
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: This field could use some `help_text`.

File: cfgov/v1/atomic_elements/organisms.py [hunk 8]
```diff
@@ -170,6 +170,46 @@ class Meta:
...
+        ('line_2', blocks.CharBlock(required=False))
+    ])
+    office_title = blocks.CharBlock(required=False)
+    more_info_url = atoms.Hyperlink(required=False)
+    positions = blocks.ListBlock(BureaStructurePosition(required=False))
+
+
+class BureaStructure(blocks.StructBlock):
+    director = blocks.CharBlock()
+    last_updated_date = blocks.DateBlock(required=False)
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: Can't we get this automatically from page metadata or something?

File: cfgov/v1/atomic_elements/organisms.py [hunk 9]
```diff
@@ -170,6 +170,46 @@ class Meta:
...
+
+
+class BureaStructureOffice(blocks.StructBlock):
+    name = blocks.CharBlock()
+    title = blocks.StructBlock([
+        ('line_1', blocks.CharBlock(required=False)),
+        ('line_2', blocks.CharBlock(required=False))
+    ])
+    office_title = blocks.CharBlock(required=False)
+    more_info_url = atoms.Hyperlink(required=False)
```
Comment (Scotchester) [inline cfgov/v1/atomic_elements/organisms.py]: This field could probably use some `help_text`.

File: cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/index.html (hunk 2 above)
Comment (sebworks) [inline cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/index.html]: Yeah, this file will be deleted once the bureau structure is migrated to Wagtail.

File: cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/_macro-content-slider.html (hunk 1 above)
Comment (sebworks) [inline cfgov/jinja2/v1/about-us/the-bureau/bureau-structure/_macro-content-slider.html]: It's getting converted to the Atomic structure.

File: cfgov/unprocessed/css/organisms/bureau-structure.less [hunk 10]
```diff
@@ -9,10 +9,10 @@
 @node-width__desktop: 405px;
 @node-width__tablet: 300px;
 @node-width__mobile: none;
-@node-border-width: 3px;
+@node-border-width: 1px;
```
Comment (sebworks) [inline cfgov/unprocessed/css/organisms/bureau-structure.less]: yeah, this should be 3px.

**Comment by schaferjh** [comment]
@sebworks Could you post a screenshot or two of what the interface for this will look like in the CMS? I'd like to get a view of what our admin users will see and verify the field labels etc. Also happy to just chat via a short demo if that's easier!

File: cfgov/jinja2/v1/_includes/organisms/bureau-structure.html [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+   _footer()
+
+   ==========================================================================
+
+   Description:
+
+   Renders an org chart footer:
+
+   ========================================================================== #}
+{% macro _footer() -%}
```
Comment (sebworks) [inline cfgov/jinja2/v1/_includes/organisms/bureau-structure.html]: It just keeps the code clean and makes it easy to make changes if necessary.

File: cfgov/jinja2/v1/_includes/organisms/bureau-structure.html [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+              o-bureau-structure_footer">
+    <div class="content-l content-l__main">
+        <div class="content-l_col content-l_col-1-2"
+             data-qa-hook="org-chart-download-info">
+            <h2 class="h3">Download a copy</h2>
+            <p>
+                Download a printable, PDF copy of this organizational chart. This version
+                is expanded to show all offices.
+            </p>
+            {# TODO Change to PDF download URL when it's available. #}
```
Comment (sebworks) [inline cfgov/jinja2/v1/_includes/organisms/bureau-structure.html]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c.

File: cfgov/jinja2/v1/_includes/organisms/bureau-structure.html [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+   value.text:     The text used within the description markup.
+
+   value.gd_code:  A GovDelivery code for a specified mailing list.
+
+   ========================================================================== #}
+<div class="o-bureau-structure">
+    <link rel="stylesheet" href="{{ static('css/slick.css')}}">
+    <h1>Bureau Structure</h1>
+    {% if value.last_updated_date %}
+    {% import 'macros/time.html' as time %}
```
Comment (sebworks) [inline cfgov/jinja2/v1/_includes/organisms/bureau-structure.html]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c.

File: cfgov/jinja2/v1/_includes/organisms/bureau-structure.html [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+    <p>
+        Org chart last updated:
+        {{ time.render( value.last_updated_date, {'date': true} ) }}
```
Comment (sebworks) [inline cfgov/jinja2/v1/_includes/organisms/bureau-structure.html]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c.

File: cfgov/v1/atomic_elements/organisms.py [hunk 15]
```diff
@@ -170,6 +170,46 @@ class Meta:
         label = ' '
 
 
+class BureaStructurePosition(blocks.StructBlock):
```
Comment (sebworks) [inline cfgov/v1/atomic_elements/organisms.py]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c.

File: cfgov/v1/atomic_elements/organisms.py (hunk 6 above)
Comment (sebworks) [inline cfgov/v1/atomic_elements/organisms.py]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c.

File: cfgov/v1/atomic_elements/organisms.py (hunk 5 above)
Comment (sebworks) [inline cfgov/v1/atomic_elements/organisms.py]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c.

File: cfgov/v1/atomic_elements/organisms.py (hunk 8 above)
Comment (sebworks) [inline cfgov/v1/atomic_elements/organisms.py]: No, the upload date may vary from any metadata that we could gather.

File: cfgov/v1/wagtail_hooks.py [hunk 16]
```diff
@@ -76,7 +76,7 @@ def editor_js():
 @hooks.register('insert_editor_css')
 def editor_css():
     css_files = [
-        'css/table-block.css',
+        'css/table-block.css', 'css/bureau-structure.css'
```
Comment (sebworks) [inline cfgov/v1/wagtail_hooks.py]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c

File: cfgov/unprocessed/css/organisms/bureau-structure.less (hunk 10 above)
Comment (sebworks) [inline cfgov/unprocessed/css/organisms/bureau-structure.less]: Fixed in https://github.com/cfpb/cfgov-refresh/pull/2532/commits/861674d42470d1ebe15c3ac204e66b4fa58f741c.

**Comment by schaferjh** [comment]
Thanks for the preview @sebworks. As promised, here are my suggested updates to the field titles: 

screen shot 2016-12-01 at 9 38 56 am

**Comment by Scotchester** [comment]
Hah, @schaferjh you countermanded [my earlier suggestion](https://github.com/cfpb/cfgov-refresh/pull/2532#discussion_r88733004) to change FROM Line 1 and Line 2 TO Title 1 and Title 2. Am I wrong about there being a situation where it isn't just for two different titles? I would not want people to view this as a formatting opportunity for people with a single title.

**Comment by schaferjh** [comment]
> I would not want people to view this as a formatting opportunity for people with a single title.

Hmmmm. I was trying to minimize the duplication of Title: Title Title, but you are right that we don't want people to start writing "Assistant" and "Director" on seperate lines just for the heck of it. 

How about this: Get rid of Title 1 and Title 2 outside the fields and keep the hint text inside the fields as Title 1 and Title 2. Getting rid of some visual clutter at least?
 
screen shot 2016-12-01 at 9 41 35 am

**Comment by sebworks** [comment]
@schaferjh, A couple notes:

- Wagtail blocks currently don't specifically support setting the placeholder text ( https://github.com/torchbox/wagtail/issues/2328 ), it only supports default text.  We can implement a workaround if you think it's necessary.

- The Hyperlink atom doesn't support default text. We can fix this issue as well if necessary.

**Comment by sebworks** [comment]
Updated screenshots:

screen shot 2016-12-05 at 11 01 59 am

screen shot 2016-12-05 at 11 01 43 am

**Comment by schaferjh** [comment]
I'm a little confused, please bear with me:

- https://github.com/torchbox/wagtail/issues/2328 seems to refer to something that we are already able to do -- In your posted screenshots, for example, the Division lead field has a label of "Division lead" and placeholder text says "Name" -- ??
  - In [my response to Scott above](https://github.com/cfpb/cfgov-refresh/pull/2532#issuecomment-264208305) I suggested removing the field labels for two text fields and just having the placeholder text -- is that not possible? If not, then I suggest "Line 1"/"Title 1" as the label and placeholder combo for these to address Scott's concern.

- Adding default text to the hyperlink was a nice to have -- no worries there. Instead could you place instructional text under that field that says "Suggested text: More information about these offices"?

- Making sure I'm reading the screen shots right: you're showing me what the admin looks like first on the Office of the Director branch and then the Divisions branch? I can assume that adding an office under a division looks the same as on the Office of the Director side, yes?

**Comment by sebworks** [comment]
> torchbox/wagtail#2328 seems to refer to something that we are already able to do -- In your posted screenshots, for example, the Division lead field has a label of "Division lead" and placeholder text says "Name" -- ??

So the screenshots show a mix of default text and placeholders. It's hard to tell the difference from a screenshot. I'll just make them placeholders.

> In my response to Scott above I suggested removing the field labels for two text fields and just having the placeholder text -- is that not possible? If not, then I suggest "Line 1"/"Title 1" as the label and placeholder combo for these to address Scott's concern.

Yep, there is always a way. Might not be pretty though.

> Adding default text to the hyperlink was a nice to have -- no worries there. Instead could you place instructional text under that field that says "Suggested text: More information about these offices"?

There currently isn't a way to provide `help text` or `default text` for the Hyperlink atom. I'll submit a PR with a fix for that. 


>  Making sure I'm reading the screen shots right: you're showing me what the admin looks like first on the Office of the Director branch and then the Divisions branch? I can assume that adding an office under a division looks the same as on the Office of the Director side, yes?

Yes, that is correct

**Comment by schaferjh** [comment]
Re: removing  the field labels for two text fields and just having the placeholder text --
If there's a "not pretty"/complicated way to do this, don't bother -- please use Line 1/Line 2 as labels and Title 1/Title 2 as placeholders.

Re: instructional (help) text under the hyperlink 
If that's not currently possible for the hyperlink atom, let's break out that change into a seperate effort and not let it hold up this PR.

Thanks!

**Comment by sebworks** [comment]
@schaferjh, fixed the labels and placeholder issues : 

screen shot 2016-12-09 at 12 04 36 pm

**Comment by schaferjh** [comment]
:+1:

**Comment by Scotchester** [comment]
Ok, some great improvements here since I last reviewed but I came across a few minor issues, still.

This screenshot shows me getting validation errors if Titles are not entered:

screen shot 2016-12-14 at 14 02 12

In the case of people like the Deputy Director and the Chief of Staff, no titles at all should be entered. In other positions, many people should only need one Title to be entered.

I think those fields need to both be optional all the time.

**Comment by chosak** [comment]
#2605 is merged. @sebworks let me know if you want me to help merge that into this PR.

**Comment by sebworks** [comment]
@chosak, merged and migration added.

**Pull request merged**